### PR TITLE
Add Python 3.8 client and common packages

### PR DESCRIPTION
**Pull request by mwperina**
Adds python38-client and python38-common packages to support oVirt
Ansibe Collection running on ansible-core-2.12 on EL8.

Signed-off-by: Martin Perina <mperina@redhat.com>

### CODE REVIEW

File: ovirt-imageio.spec.in [hunk 1]
```diff
@@ -13,17 +13,41 @@ Source0:   https://github.com/oVirt/%{name}/releases/download/v%{version}/%{name
 %global admin_confdir %{_sysconfdir}/%{name}
 %global vendor_confdir %{_prefix}/lib/%{name}
 
+# We need to have Python 3.8 support of imageio client for oVirt Ansible
+# Collection running with ansible-core-2.12, which uses non-standard Python 3.8
+# on EL8 platform
+%if "0%{?rhel}" == "08"
```
Comment (nirs) [inline ovirt-imageio.spec.in:19]: Why compare to "08" instead of "8"?

Comment (mnecas) [inline ovirt-imageio.spec.in:19]: Shouldn't the check be `%if 0%{?rhel} > 7 && 0%{?rhel} < 9`, or `%if "0%{?rhel}" == 8`?

Comment (mwperina) [reply ovirt-imageio.spec.in:19]: I hate `%if 0%{?rhel} > 7 && 0%{?rhel} < 9` and `%if 0%{?rhel} == 8` never worked for me, so trying a new variant, let's see :-)

File: ovirt-imageio.spec.in [hunk 2]
```diff
@@ -145,6 +172,53 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
+%{python3_sitearch}/%{srcname}-*.egg-info
+%exclude %{python3_sitearch}/%{srcname}/client
+%exclude %{python3_sitearch}/%{srcname}/admin
+
+
+%package python38-client
+Summary:   oVirt imageio client library
+
+# NOTE: keep in sync with automation/check.packages
+Requires:  python38-%{name}-common = %{version}-%{release}
```
Comment (mnecas) [inline ovirt-imageio.spec.in:204]: This requires the `python38-ovirt-imageio-common` but above it is build with name `ovirt-imageio-python38-common`, please change the requirement name.

**Review by nirs — CHANGES_REQUESTED** [review]

File: ovirt-imageio.spec.in (hunk 1 above)
Comment (mnecas) [reply ovirt-imageio.spec.in:19]: Interesting I had no problem running `%if 0%{?rhel} == 8` inside my el8 container.

File: ovirt-imageio.spec.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwperina) [reply ovirt-imageio.spec.in]: Right, fixed

File: ovirt-imageio.spec.in [hunk 3]
```diff
@@ -145,6 +172,53 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
+Summary:   oVirt imageio common resources
+
+# NOTE: keep in sync with automation/build-artifacts.packages
+BuildRequires: gcc
+BuildRequires: python38-devel
+
+# NOTE: keep in sync with automation/check.packages
+Requires:  python38
+
+%description python38-imageio-common
```
Comment (mnecas) [inline ovirt-imageio.spec.in:189]: Please change to `%description -n python38-imageio-common` without the -n it tries to make `ovirt-imageio-python38-imageio-common`

File: ovirt-imageio.spec.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (nirs) [reply ovirt-imageio.spec.in]: But I'm ok with having standard packages names for the python 3.8 varaints:

- python38-ovirt-imageio-common
- python38-ovirt-imageio-client

I'm not why we did not move to this naming when moving to python 3.

File: ovirt-imageio.spec.in (hunk 3 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Done

File: docker/Dockerfile.centos-8 [hunk 4]
```diff
@@ -16,6 +16,7 @@ RUN echo v1 \
         make \
         openssl \
         python3-devel \
+        python38-devel \
```
[the remark this answers is not in the document]
Comment (mwperina) [reply docker/Dockerfile.centos-8]: Done

File: ovirt-imageio.spec.in (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwperina) [reply ovirt-imageio.spec.in:204]: Hopefully fixed now

File: Makefile [hunk 5]
```diff
@@ -29,6 +29,7 @@ srpm: dist
 		-ts "$(OUTDIR)/$(TAR_NAME)"
 
 rpm: srpm
+	dnf builddep -y "$(OUTDIR)/$(PACKAGE_NAME)-$(PACKAGE_VERSION)"*.src.rpm
```
[the remark this answers is not in the document]
Comment (mwperina) [reply Makefile]: Removed

File: ovirt-imageio.spec.in (hunk 1 above)
Comment (mwperina) [reply ovirt-imageio.spec.in:19]: We have tried in other projects `%if 0%{?rhel} == 8` and it didn't work, so I've tried to compare strings and it works now. But maybe it could work even without 0

**Comment by mwperina** [comment]
Generated RPMS looks good to me, but there is a difference in requires which I cannot explain:

```
$ rpm -qp --requires python38-ovirt-imageio-common-2.4.3-0.202204011513.gitd0b0d82.el8.x86_64.rpm
libc.so.6()(64bit)
libc.so.6(GLIBC_2.10)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.4)(64bit)
libpthread.so.0()(64bit)
python(abi) = 3.8
python38
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PartialHardlinkSets) <= 4.0.4-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1
rtld(GNU_HASH)
```

```
$ rpm -qp --requires ovirt-imageio-common-2.4.3-0.202204011513.gitd0b0d82.el8.x86_64.rpm
libc.so.6()(64bit)
libc.so.6(GLIBC_2.10)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.4)(64bit)
libpthread.so.0()(64bit)
libpython3.6m.so.1.0()(64bit)
python(abi) = 3.6
python3
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PartialHardlinkSets) <= 4.0.4-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsXz) <= 5.2-1
rtld(GNU_HASH)
```

And similar difference in provides:

```
$ rpm -qp --provides python38-ovirt-imageio-common-2.4.3-0.202204011513.gitd0b0d82.el8.x86_64.rpm
python3.8dist(ovirt-imageio) = 2.4.3
python38-ovirt-imageio-common = 2.4.3-0.202204011513.gitd0b0d82.el8
python38-ovirt-imageio-common(x86-64) = 2.4.3-0.202204011513.gitd0b0d82.el8
```

```
$ rpm -qp --provides ovirt-imageio-common-2.4.3-0.202204011513.gitd0b0d82.el8.x86_64.rpm
ovirt-imageio-common = 2.4.3-0.202204011513.gitd0b0d82.el8
ovirt-imageio-common(x86-64) = 2.4.3-0.202204011513.gitd0b0d82.el8
python3.6dist(ovirt-imageio) = 2.4.3
python3dist(ovirt-imageio) = 2.4.3
```

File: ovirt-imageio.spec.in (hunk 1 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Fixed

**Comment by tinez** [comment]
> Generated RPMS looks good to me, but there is a difference in requires which I cannot explain:

They don't work for me - imageio daemon seems to be using python3.8 shebangs:

```
[root@ost-basic-suite-master-engine ~]# head -n1 /usr/bin/ovirt-imageio
#!/usr/bin/python3.8 -s
```

Also, the python3.8 versions of the RPMs weren't installed for me when I added the repo from the actions run. What should enforce the installation of the other versions?

**Comment by mwperina** [comment]
> > Generated RPMS looks good to me, but there is a difference in requires which I cannot explain:
> 
> They don't work for me - imageio daemon seems to be using python3.8 shebangs:

Hmm, python 3.8 build is executed after 3.6, so it will overwrite shared files. I've just reordered so 3.8 is first and only afterwards 3.6, because we don't need 3.8 daemon.

> 
> Also, the python3.8 versions of the RPMs weren't installed for me when I added the repo from the actions run. What should enforce the installation of the py38 versions?

https://github.com/oVirt/ovirt-ansible-collection/pull/467

**Comment by tinez** [comment]
> Hmm, python 3.8 build is executed after 3.6, so it will overwrite shared files. I've just reordered so 3.8 is first and only afterwards 3.6, because we don't need 3.8 daemon.

I was about to suggest the same :+1:

**Comment by tinez** [comment]
It didn't work unfortunately

**Comment by mwperina** [comment]
It's ugly, but now we have correct shebang in the daemon package. So the only somewhat problematic issue is that `python38-ovirt-imageio-common` doesn't have requires for `libpython3.8m.so.1.0()(64bit)`, but not sure if that's really needed or not

File: ovirt-imageio.spec.in [hunk 6]
```diff
@@ -94,8 +132,9 @@ Daemon providing image transfer service on oVirt hosts.
 
 %files daemon
 %{python3_sitearch}/%{srcname}/admin
-%{_bindir}/%{name}
-%{_bindir}/%{name}ctl
+# Marking below
+%attr(755, root, root) %{_bindir}/%{name}
+%attr(755, root, root) %{_bindir}/%{name}ctl
```
Comment (nirs) [inline ovirt-imageio.spec.in:137]: Why is this needed? Isn't this the default?

File: ovirt-imageio.spec.in [hunk 7]
```diff
@@ -145,6 +184,56 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
 fi
 
+%if %{with_python38}
+%define python3_pkgversion 38
+%define __python3 /usr/bin/python3.8
+
+%package -n python38-%{name}-common
+Summary:   oVirt imageio common resources
+
+# NOTE: keep in sync with automation/build-artifacts.packages
```
Comment (nirs) [inline ovirt-imageio.spec.in:194]: This comment is not needed now, no such file now.

File: ovirt-imageio.spec.in [hunk 8]
```diff
@@ -145,6 +184,56 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
+BuildRequires: gcc
+BuildRequires: python38-devel
+
```
Comment (nirs) [inline ovirt-imageio.spec.in:197]: Any reason for the blank line here?

File: ovirt-imageio.spec.in [hunk 9]
```diff
@@ -145,6 +184,56 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
+BuildRequires: sed
+
+# NOTE: keep in sync with automation/check.packages
```
Comment (nirs) [inline ovirt-imageio.spec.in:200]: comment not need now.

File: ovirt-imageio.spec.in [hunk 10]
```diff
@@ -145,6 +184,56 @@ if systemctl is-active ovirt-imageio-daemon.service >/dev/null; then
...
+%{python3_sitearch}/%{srcname}
+%{python3_sitearch}/%{srcname}-*.egg-info
+%exclude %{python3_sitearch}/%{srcname}/client
+%exclude %{python3_sitearch}/%{srcname}/admin
+
+
+%package -n python38-%{name}-client
+Summary:   oVirt imageio client library
+
+# NOTE: keep in sync with automation/check.packages
```
Comment (nirs) [inline ovirt-imageio.spec.in:217]: Not needed

File: ovirt-imageio.spec.in [hunk 11]
```diff
@@ -32,10 +56,24 @@ install -D -m 0755 --directory %{buildroot}%{admin_confdir}/conf.d
...
 
+%if %{with_python38}
+# Python 3.8 wins shebangs mangling, but we want daemon to be available only
+# for python 3, so we need to override
+mv %{buildroot}%{_bindir}/%{name} %{buildroot}%{_bindir}/%{name}.bck
+mv %{buildroot}%{_bindir}/%{name}ctl %{buildroot}%{_bindir}/%{name}ctl.bck
+sed -e 's|python3.8|python3|g' %{buildroot}%{_bindir}/%{name}.bck > %{buildroot}%{_bindir}/%{name}
+sed -e 's|python3.8|python3|g' %{buildroot}%{_bindir}/%{name}ctl.bck > %{buildroot}%{_bindir}/%{name}ctl
+rm %{buildroot}%{_bindir}/%{name}.bck
+rm %{buildroot}%{_bindir}/%{name}ctl.bck
```
Comment (nirs) [inline ovirt-imageio.spec.in:67]: This is horrible hack :-)

Why not use sed -i to change the file inplace?

**Review by nirs — CHANGES_REQUESTED** [review]

**Comment by nirs** [comment]
Please also rebase

**Comment by nirs** [comment]
> It's ugly, but now we have correct shebang in the daemon package. So the only somewhat problematic issue is that `python38-ovirt-imageio-common` doesn't have requires for `libpython3.8m.so.1.0()(64bit)`, but not sure if that's really needed or not

The tests pass, so we know it works with current python 3.8 package.

File: ovirt-imageio.spec.in (hunk 6 above)
Comment (michalskrivanek) [reply ovirt-imageio.spec.in:137]: I guess because of those ">" above. Indeed we should just sed -i instead

File: ovirt-imageio.spec.in (hunk 11 above)
Comment (mwperina) [reply ovirt-imageio.spec.in:67]: Yes, it is. The problem is that shebang mangling is performed for executable files, so we need fix shebang here, then make the file nonexecutable and then only during RPM installation make it executable again (line 136-137).

But Marcin found out better solution, I will post it soon

File: ovirt-imageio.spec.in (hunk 6 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Not needed anymore, Marcin's solution is better

File: ovirt-imageio.spec.in (hunk 9 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Removed

File: ovirt-imageio.spec.in (hunk 10 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Removed

File: ovirt-imageio.spec.in (hunk 11 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Not needed anymore, Marcin's solution is better and works fine

File: ovirt-imageio.spec.in (hunk 7 above)
Comment (mwperina) [reply ovirt-imageio.spec.in]: Removed

File: ovirt-imageio.spec.in (hunk 8 above)
Comment (mwperina) [reply ovirt-imageio.spec.in:179]: Isn't more readable to have a blank line between BuildRequires and Requires?

Comment (nirs) [reply ovirt-imageio.spec.in:179]: Right, I missed it.

**Comment by nirs** [comment]
Tested on RHEL 8.6 by installing the pythhon38 packages and uploading and downloading
using python3.8.

```
$ python3.8 /usr/share/doc/python38-ovirt-engine-sdk4/examples/upload_disk.py -c engine --sd-name alpine-iscsi-02 --disk-sparse --disk-format qcow2 /var/tmp/fedora-35.qcow2 
[   0.0 ] Checking image...
[   0.0 ] Image format: qcow2
[   0.0 ] Disk format: cow
[   0.0 ] Disk content type: data
[   0.0 ] Disk provisioned size: 6442450944
[   0.0 ] Disk initial size: 1219100672
[   0.0 ] Disk name: fedora-35.qcow2
[   0.0 ] Disk backup: False
[   0.0 ] Connecting...
[   0.0 ] Creating disk...
[   4.5 ] Disk ID: b657faf0-8e7b-41ff-affa-64df0aadb438
[   4.5 ] Creating image transfer...
[   7.1 ] Transfer ID: 959559e3-2b73-4848-b333-663a4342023d
[   7.1 ] Transfer host name: host4
[   7.1 ] Uploading image...
[ 100.00% ] 6.00 GiB, 3.17 seconds, 1.89 GiB/s                                 
[  10.2 ] Finalizing image transfer...
[  14.3 ] Upload completed successfully

$ python3.8 /usr/share/doc/python38-ovirt-engine-sdk4/examples/download_disk.py -c engine b657faf0-8e7b-41ff-affa-64df0aadb438 /var/tmp/download.qcow2
[   0.0 ] Connecting...
[   0.2 ] Creating image transfer...
[   2.7 ] Transfer ID: 70213605-b0d5-48f0-9fc2-c03700e7d7ec
[   2.7 ] Transfer host name: host4
[   2.7 ] Downloading disk...
[ 100.00% ] 6.00 GiB, 2.03 seconds, 2.95 GiB/s                                 
[   4.8 ] Finalizing image transfer...
[   9.9 ] Download completed successfully

$ qemu-img compare -s /var/tmp/fedora-35.qcow2 /var/tmp/download.qcow2 
Images are identical.
```

**Review by nirs — APPROVED** [review]

**Review by nirs — APPROVED** [review]

**Review by bennyz — APPROVED** [review]

**Review by barpavel — COMMENTED** [review]
LGTM

File: ovirt-imageio.spec.in [hunk 12]
```diff
@@ -81,9 +107,9 @@ Python client library for accessing imageio server on oVirt hosts.
 %package daemon
 Summary:   oVirt imageio daemon
 
-# NOTE: keep in sync with docs/development.md
 BuildRequires: systemd
 
+Requires:  %{name}-common = %{version}-%{release}
```
[the remark this answers is not in the document]
Comment (nirs) [reply ovirt-imageio.spec.in:112]: Bad reabse, fixed in #63
